rakelib/examples.rake

```diff
@@ -0,0 +1,107 @@
+# Defines tasks related to generating example apps using the gem's generator.
+# Allows us to create and test apps generated using a wide range of options.
+#
+# Also see example_type.rb
+require "yaml"
+
+require_relative "example_type"
+require_relative "task_helpers"
+
+include ReactOnRails::TaskHelpers
+
+namespace :examples do
+  # Loads data from examples_config.yml and instantiates corresponding ExampleType objects
+  examples_config_file = File.expand_path("../examples_config.yml", __FILE__)
+  examples_config = YAML.load(File.read(examples_config_file))
+  examples_config["example_type_data"].each do |example_type_data|
+    example_type_data.keys.each { |key| example_type_data[key.to_sym] = example_type_data.delete(key) }
+    ExampleType.new(example_type_data)
+  end
+
+  # Define tasks for each example type
+  ExampleType.all.each do |example_type|
+    # GENERATED FILES
+    example_type.generated_files.each do |f|
+      file f => example_type.source_files do
```

Comment:
this is an example of a `file` task in Rake. We are using a FileList of all generated files for the given example type, and declaring them dependent on a FileList of all of the source files (in lib/generators/react_on_rails). This dependency means that if the dependee has a newer modification date than then dependent, then the block will be invoked. In this case, we invoke a task to re-generate the app. Note that rake keeps track of tasks being invoked, and therefore will not invoke the same task more than twice.